checkstyle-tester/projects-to-test-on-for-github-action.properties

```diff
@@ -38,7 +38,7 @@ Vavr|git|https://github.com/vavr-io/vavr|v0.9.0||
 # custom javadoc tags
 nbia-dcm4che-tools|git|https://github.com/thprakash/nbia-dcm4che-tools|c3591e6f0f84827586db25abded6708e5386ef1a||
 # RequireThis usage
-spring-integration|git|https://github.com/spring-projects/spring-integration.git|main|
+spring-integration|git|https://github.com/spring-projects/spring-integration.git|origin/main|
```

Comment:
I do not see a good reason to do this; if anything, not specifying a remote here helped to figure out that local repo had bad git settings (wrong default branch).

IMO, this PR and https://github.com/checkstyle/contribution/pull/761 show that we have deficiencies in the launch_diff.sh script, not that we need to modify the project files.

Reply:
And this also shows nearly everything that is deficient in `launch_diff.sh` is also deficient in `diff.groovy`. `main` is a local branch name. This is relying on the fact the clone and reset method, which both scripts only do, creates a local copy which it didn't do in this case since the repo never had it originally. There is nothing "bad" about the git copies I have besides that they are old clones and we don't support updating them ( https://github.com/checkstyle/contribution/issues/180 ).


````
transfers spring integration to local

PS Z:\contribution\checkstyle-tester> type .\projects-to-test-on.properties
spring-integration|git|https://github.com/spring-projects/spring-integration.git|main|

PS Z:\contribution\checkstyle-tester> groovy .\diff.groovy -r M:\checkstyleWorkspaceEclipse\checkstyle -p patch -l projects-to-test-on.properties -c M:\checkstyleWorkspaceEclipse\contribution\checkstyle-tester\projects-to-test-on.properties
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.codehaus.groovy.reflection.CachedClass (file:/C:/RickysPrograms/groovy-2.4.8/lib/groovy-2.4.8.jar) to method java.lang.Object.finalize()
WARNING: Please consider reporting this to the maintainers of org.codehaus.groovy.reflection.CachedClass
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
   [delete] Deleting directory Z:\contribution\checkstyle-tester\reports
Installing Checkstyle artifact (master) into local Maven repository ...
Running command: git checkout master
Already on 'master'Your branch is up to date with 'origin/master'.

Running command: git log -1 --pretty=MSG:%s%nSHA-1:%H
MSG:Issue #12589: resolves naming pitest suppressions
SHA-1:92e9abf3a23edd1d4a24a30e672104cc8761fb3f
Running command: mvn -e --no-transfer-progress --batch-mode -Pno-validations clean install
....
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  47.749 s
[INFO] Finished at: 2023-01-06T08:57:43-05:00
[INFO] ------------------------------------------------------------------------
Testing Checkstyle started
   [delete] Deleting directory Z:\contribution\checkstyle-tester\src\main\java
Resetting git sources to commit 'main'
Running command: git reset --hard main
fatal: ambiguous argument 'main': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
Caught: groovy.lang.GroovyRuntimeException: Error: !
groovy.lang.GroovyRuntimeException: Error: !
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at diff.executeCmd(diff.groovy:686)
        at diff$executeCmd$8.callCurrent(Unknown Source)
        at diff.cloneRepository(diff.groovy:375)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at diff$_generateCheckstyleReport_closure4.doCall(diff.groovy:294)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at diff.generateCheckstyleReport(diff.groovy:270)
        at diff$generateCheckstyleReport$11.callCurrent(Unknown Source)
        at diff.launchCheckstyleReport(diff.groovy:231)
        at diff$launchCheckstyleReport$7.callCurrent(Unknown Source)
        at diff.run(diff.groovy:31)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
````

There are other instances where `diff.groovy` relies too much on local branches when it should be using remote. One example is the default regression process uses `master`, a local branch, and not the remote branch ( https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/diff.groovy#L788 ). Locals are possibly a user's creation of the remote branch, can be modified, or not even checked out because it changes too much. It should also be using the remote branch ( https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/launch_diff.sh#L278 ).

````
PS M:\checkstyleWorkspaceEclipse\contribution\checkstyle-tester> groovy .\diff.groovy -r M:\checkstyleWorkspaceEclipse\checkstyle -p patch -l projects-to-test-on.properties -c M:\checkstyleWorkspaceEclipse\contribution\checkstyle-tester\projects-to-test-on.properties
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.codehaus.groovy.reflection.CachedClass (file:/C:/RickysPrograms/groovy-2.4.8/lib/groovy-2.4.8.jar) to method java.lang.Object.finalize()
WARNING: Please consider reporting this to the maintainers of org.codehaus.groovy.reflection.CachedClass
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
   [delete] Deleting directory M:\checkstyleWorkspaceEclipse\contribution\checkstyle-tester\reports
Installing Checkstyle artifact (master) into local Maven repository ...
Running command: git checkout master
error: pathspec 'master' did not match any file(s) known to git
hint: 'master' matched more than one remote tracking branch.
hint: We found 11 remotes with a reference that matched. So we fell back
hint: on trying to resolve the argument as a path, but failed there too!
hint:
hint: If you meant to check out a remote tracking branch on, e.g. 'origin',
hint: you can do so by fully qualifying the name with the --track option:
hint:
hint:     git checkout --track origin/<name>
hint:
hint: If you'd like to always have checkouts of an ambiguous <name> prefer
hint: one remote, e.g. the 'origin' remote, consider setting
hint: checkout.defaultRemote=origin in your config.
Caught: groovy.lang.GroovyRuntimeException: Error: !
groovy.lang.GroovyRuntimeException: Error: !
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at diff.executeCmd(diff.groovy:686)
        at diff$executeCmd$8.callCurrent(Unknown Source)
        at diff.launchCheckstyleReport(diff.groovy:222)
        at diff$launchCheckstyleReport$7.callCurrent(Unknown Source)
        at diff.run(diff.groovy:31)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
````

Reply:
>There is nothing "bad" about the git copies I have besides that they are old clones and we don't support updating them

As I mention in https://github.com/checkstyle/contribution/pull/760#discussion_r1063438329, the clone is not bad, the git settings in the clone are.

If we want to specify remote branches, ok, but then we should do it for any project that specifies a branch instead of a commit SHA.

```
➜  checkstyle-tester git:(master) ✗ ag -G '\.properties' 'master|main' . | wc -l
18

```

Reply:
>Locals are possibly a user's creation of the remote branch, can be modified, or not even checked out because it changes too much. It should also be using the remote branch...

This is a good point. We could probably make a bunch of improvements in all our scripts that rely on git, like single depth cloning a remote branch only.

Reply:
> we should do it for any project that specifies a branch instead of a commit SHA.

Yes we should.

> like single depth cloning a remote branch only

Shallow clones will make updating worse unless there is some workaround since we only do clones and resets and not updates ( https://github.com/checkstyle/contribution/issues/351 ).